### PR TITLE
feat(sdk-core): dkls signing with wallet platform

### PR DESCRIPTION
**Pull request by pranavjain97**
TICKET: WP-1680

### CODE REVIEW

File: modules/sdk-core/src/bitgo/tss/ecdsa/dkls-types.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import * as t from 'io-ts';
+import { NonEmptyString } from 'io-ts-types';
+import { createStringObject } from '@bitgo-private/common-interface';
+
+/**
+ * Note these types are temporarily till we move types to wallet-platform-types from WP.
+ * Ignore any comiple changes on this file
+ */
+
+export const MpcPartyEnum = createStringObject(['1', '2', '3']);
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/tss/ecdsa/dkls-types.ts:10]:
```suggestion
export const MpcPartyEnum = createStringObject(['0', '1', '2']);
```

File: modules/sdk-core/src/bitgo/tss/ecdsa/dkls-types.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
+import * as t from 'io-ts';
+import { NonEmptyString } from 'io-ts-types';
+import { createStringObject } from '@bitgo-private/common-interface';
+
+/**
+ * Note these types are temporarily till we move types to wallet-platform-types from WP.
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/tss/ecdsa/dkls-types.ts:6]: https://github.com/BitGo/wallet-platform-types/blob/master/src/schema/keyGeneration/MPCv2/messages.ts maybe these can be used and moved to a more generic location

File: modules/sdk-core/src/bitgo/tss/ecdsa/dkls.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+  const userToBitGoEncryptedMsg3 = await DklsComms.encryptAndAuthOutgoingMessages(
+    DklsTypes.serializeMessages(userToBitGoMessages3),
+    [getBitGoPartyGpgKey(bitgoGpgKey)],
+    [getUserPartyGpgKey(userGpgKey)]
+  );
+  assert(MPCv2PartyFromStringOrNumber.is(userToBitGoEncryptedMsg2.p2pMessages[0].from));
+  assert(MPCv2PartyFromStringOrNumber.is(userToBitGoEncryptedMsg2.p2pMessages[0].to));
+  assert(MPCv2PartyFromStringOrNumber.is(userToBitGoEncryptedMsg3.p2pMessages[0].from));
+  assert(MPCv2PartyFromStringOrNumber.is(userToBitGoEncryptedMsg3.p2pMessages[0].to));
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/tss/ecdsa/dkls.ts:73]: wont all these throw a `cannot access foo from undefined `if the array is empty?

Comment (alebusse) [reply modules/sdk-core/src/bitgo/tss/ecdsa/dkls.ts:73]: maybe assert that exists and use a codec to check the entire msgs?

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts [hunk 4]
```diff
@@ -473,6 +487,156 @@ export class EcdsaMPCv2Utils extends BaseEcdsaUtils {
...
+    if (requestType === RequestType.tx) {
+      assert(txRequest.transactions || txRequest.unsignedTxs, 'Unable to find transactions in txRequest');
+      const unsignedTx =
+        txRequest.apiVersion === 'full' ? txRequest.transactions![0].unsignedTx : txRequest.unsignedTxs[0];
+      msgToSign = unsignedTx.signableHex;
+      derivationPath = unsignedTx.derivationPath;
+    } else if (requestType === RequestType.message) {
+      throw new Error('DKLS message signing not supported yet.');
+    }
+
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts:526]:
```suggestion
    } else { throw }

```

Comment (alebusse) [reply modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts:526]: Not sure what other type can be, it must be `tx`

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts [hunk 5]
```diff
@@ -473,6 +487,156 @@ export class EcdsaMPCv2Utils extends BaseEcdsaUtils {
...
+    if (requestType === RequestType.tx) {
+      assert(txRequest.transactions || txRequest.unsignedTxs, 'Unable to find transactions in txRequest');
+      const unsignedTx =
+        txRequest.apiVersion === 'full' ? txRequest.transactions![0].unsignedTx : txRequest.unsignedTxs[0];
+      msgToSign = unsignedTx.signableHex;
+      derivationPath = unsignedTx.derivationPath;
+    } else if (requestType === RequestType.message) {
+      throw new Error('DKLS message signing not supported yet.');
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts:524]:
```suggestion
      throw new Error('MPCv2 message signing not supported yet.');
```

File: modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts [hunk 6]
```diff
@@ -473,6 +487,156 @@ export class EcdsaMPCv2Utils extends BaseEcdsaUtils {
...
+
+  private async signRequestBase(params: TSSParams | TSSParamsForMessage, requestType: RequestType): Promise<TxRequest> {
+    const userKeyShare = Buffer.from(params.prv, 'base64');
+    const txRequest: TxRequest =
+      typeof params.txRequest === 'string'
+        ? await getTxRequest(this.bitgo, this.wallet.id(), params.txRequest)
+        : params.txRequest;
+
+    let derivationPath = '';
+    let msgToSign = '';
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/utils/tss/ecdsa/ecdsaMPCv2.ts:510]: Im not sure about initializing this as an empty string, is a valid value even if you dont get it from the tx itself

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 7]
```diff
@@ -141,7 +141,11 @@ export class Wallet implements IWallet {
     if (baseCoin?.supportsTss() && this._wallet.multisigType === 'tss') {
       switch (baseCoin.getMPCAlgorithm()) {
         case 'ecdsa':
-          this.tssUtils = new EcdsaUtils(bitgo, baseCoin, this);
+          if (walletData.multisigTypeVersion === 'MpcV2') {
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:144]:
```suggestion
          if (walletData.multisigTypeVersion === 'MPCv2') {
```
Lets add it to the public-types too

File: modules/sdk-core/src/bitgo/wallet/wallet.ts [hunk 8]
```diff
@@ -141,7 +141,11 @@ export class Wallet implements IWallet {
...
+            this.tssUtils = new EcdsaUtils(bitgo, baseCoin, this);
+          } else {
+            this.tssUtils = new EcdsaMPCv2Utils(bitgo, baseCoin, this);
```
Comment (alebusse) [inline modules/sdk-core/src/bitgo/wallet/wallet.ts:147]:
```suggestion
          this.tssUtils = new EcdsaMPCv2Utils(bitgo, baseCoin, this);
          } else {
           this.tssUtils = new EcdsaUtils(bitgo, baseCoin, this);
```
wrong order

File: yarn.lock
Comment (alebusse) [inline yarn.lock:1]: run git clean -fdx and yarn install too see if this yarn.lock change suppose to be here

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 8 above)
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:147]: nice catch!

File: modules/sdk-core/src/bitgo/tss/ecdsa/dkls.ts (hunk 3 above)
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/tss/ecdsa/dkls.ts:73]: added a check to see the length before. I have a TODO to decode the API messages using codec, i tried that but for some reason the decode gives an error even though the response matches the codec. I'm gonna do this as a follow-up.

File: modules/sdk-core/src/bitgo/wallet/wallet.ts (hunk 7 above)
Comment (pranavjain97) [reply modules/sdk-core/src/bitgo/wallet/wallet.ts:144]: Let's do that as a follow-up?

File: yarn.lock
Comment (pranavjain97) [reply yarn.lock:1]: yeah its the same

**Review by mkottaichamy — APPROVED** [review]

**Pull request merged**